portfolio/src/main/java/com/google/sps/servlets/CommentsDatastore.java

```diff
@@ -0,0 +1,70 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.List;
+import java.util.ArrayList;
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+public class CommentsDatastore {
+
+  public class Comment {
+    private String comment;
+    private long timestamp;
+
+    public Comment(String comment, long timestamp) {
+      this.comment = comment;
+      this.timestamp = timestamp;
+    }
+  } 
+
+    public CommentsDatastore() {}
+    
+    public List<Comment> fetchComments(int maxNumOfComments) {
+      Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      List<Entity> results = datastore.prepare(query)
+        .asList(FetchOptions.Builder.withLimit(maxNumOfComments));
+      
+      List<Comment> comments = new ArrayList<>();
+      for (Entity entity : results) {
+        String comment = (String) entity.getProperty("comment");
+        long timestamp = (long) entity.getProperty("timestamp");
+        Comment newComment = new Comment(comment, timestamp);
+        comments.add(newComment);
+      }
+
+      return comments;
+    }
+
+    public Comment addComment(String comment, int maxNumOfComments) {
+      long timestamp = System.currentTimeMillis();
+
+      if (comment != null && !comment.isEmpty()) {
```

Comment:
comment is not nullable. Anything not marked with @Nullable is assumed non null. Also you probably don't want it nullable anyway. Empty string is a valid check though. You probably want to throw IllegalArgumentException if comment is empty

Reply:
I think I had to check if comments was null because I call the same servlet when I change the maxNumOfComments. Do you think I should separate out the servlets? I just though it'd be easier to check if comments/maxNumOfComments was null and handle them both in the same function.

Reply:
your maxNumComments is not stored anywhere though and not related to "addComment" where you create a comment and add it to the datastore. i'm pretty sure this variable is unused right now.